utils.bash

```diff
@@ -0,0 +1,147 @@
+# Not an executable. Can be sourced by other scripts in this repo for some free
+# error checking for common operations. Uses bash-specific features including
+# `local`.
+
+# see bash(1)
+function set_strict_mode {
+  set -euxo pipefail
+}
+
+# Display a message to stderr.
+function warn {
+  echo >&2 "$@"
+}
+
+# Display a message to stderr, then exit with failure.
+function die {
+  echo >&2 "$@"
+  exit 1
+}
+
+# Display a message read from stdin to stderr, then exit with failure. This can
+# be used with heredocs (<<EOF and friends).
+function die_here {
+  cat >&2
+  exit 1
+}
+
+# Check for the existence of a command, and fail if it's not there. If a command
+# has variations in functionality between Linux and OSX, additional checks may
+# be required.
+function check_cmd_or_err {
+  local -r cmd_name="$1"
+  if ! hash "$cmd_name"; then
+    die_here <<EOF
+The command '${cmd_name}' is required to run this script. You may have to
+install it using your operating system's package manager.
+EOF
+  fi
+}
+
+# Verify that a path exists, and echo the absolute path (not canonical or
+# normalized!).
+function get_existing_absolute_path {
+  local -r path_arg="$1"
+  # -f is an "illegal option" for OSX readlink, so this may have e.g. `/../`
+  # within it.
+  local -r abs_path="$(pwd)/${path_arg}"
+
+  if [[ ! -e "$abs_path" ]]; then
+    die "File at path '${path_arg}' (relative to pwd '$(pwd)') was expected to exist, but does not."
+  fi
+
+  echo "$abs_path"
+}
+
+function var_is_set {
+  local -r var_name="$1"
+  [[ "${!var_name+x}" != '' ]]
```

Comment:
What's the `+x` here?

(Also, this is perfectly fine and readable, but `-z` (empty string) and `-n` (non-empty) are handy tests to know about :))